libsmi/tools/smiquery.1
.\"
.\" $Id: smiquery.1.in 1174 2004-08-10 10:58:12Z strauss $
.\"
.TH smiquery 1  "August 10, 2004" "IBR" "SMI Tools"
.SH NAME
smiquery \- query single information from SMI MIB modules
.SH SYNOPSIS
.B smiquery
[
.B "-Vh"
] [
.BI "-p " module
]
.I command name
.SH DESCRIPTION
The \fBsmiquery\fP program is used to query information on a single
item from an MIB module, e.g. module meta information or a single type
definition.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smiquery version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit. The help text lists all available \fBcommand\fPs.
.TP
\fB-p \fImodule\fB, --preload \fImodule\fP
Preload the module \fImodule\fP before reading the main module. This may
be helpful if an incomplete main module misses to import some definitions.
.TP
.I command name
This specifies the kind of information to retrieve (module, imports,
revisions, node, parent, compliance, index, members, children, type, or
macro) and the item to retrieve.
.PP
The module in question is searched along a path, which is initialized by
the \fBSMIPATH\fP environment variable or by /Users/schoenw/share/mibs if
\fBSMIPATH\fP is not set. Files with a \fB.sming\fP extension are
expected to be (old-style!) SMIng module files. Files
with a \fB.my\fP or without
any extension are expected to be SMIv1/v2 module files.
.SH "EXAMPLE"
This example queries for information on the node ifNumber defined
in the module IF-MIB.
.nf

  $ smiquery node IF-MIB::ifNumber
       MibNode: IF-MIB::ifNumber
	   OID: 1.3.6.1.2.1.2.1
	  Type: Integer32
   Declaration: OBJECT-TYPE
      NodeKind: scalar
	Access: read-only
	Status: current
   Description: The number of network interfaces (regardless of their
		current state) present on this system.

.fi
.SH "SEE ALSO"
The
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
.SH "AUTHOR"
(C) 1999-2004 F. Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br


qwt/CHANGES
Release 5.1.0
===================

Changes
-------
1) QwtSymbol::copy introduced
   Now it is possible to use derived symbol classes for curves
2) QwtPlotScaleItem introduced
   A new type of plot item for displaying axes on the canvas
3) QwtClipper added
   A collection of clipping algos
4) Using DBL_EPSILON
   This change allows smaller intervals for sliders/dials
5) QwtPanner
   setOrientation() added.
6) QwtPlot
   axisStepSize() added
   clear is virtual now
7) QwtPlotPrintFilter
   PrintCanvasBackground splitted into PrintBackground, PrintFrameWithScales
8) QwtPlotZoomer
   setZoomStack() added
9) Changes for the QwtPolar package
   QwtLegendItemManager introduced
   QwtMagnifier introduced
10)Suffix rules added in qwtconfig.pri for different targets for 
   debug/release builds.

Bug Fixes
---------
1. QwtAbstractScaleDraw::setAbstractScaleDraw
   Reinitialization problem fixed
2. QwtLegendItem
   key event handlers fixed
3. QwtPicker
   solaris-cc compiler problem fixed
4. Inaccurate mapping of scale to widget coordinates fixed
5. QwtPlotCurve::draw
   Updates for Qt 4.3 added
6. QwtPlotLayout
   AlignToCanvas layout calculation fixed
7. QwtPlot::print
   Workaround for a QPen initialization problem, 
   when printing to Pdf, added
8. QwtText
   Layout of rich text documents fixed
9. Designer
   Handling of QwtScaleWidget fixed
10. realtime example
   Qt::WA_PaintOutsidePaintEvent added, ScrollZoomer fixed
11. Several others I have forgotten

Release 5.0.2
===================

Bug Fixes
---------
1. QwtPlotCurve::Xfy curve type fixed
2. Memory leak in QwtLegend fixed
3. Vertical alignment of rich texts fixed
4. Workaround for a Qt4 bug added, that produces horrible performance
   when painting curves with a pen width > 1.
5. Background for the tracker text of QwtPickers fixed.
   Improved (faster + better rendered texts) implementation of
   painting tracker texts, using capabilities of Qt >= 4.3.
6. QwtArrowButton/QwtCounter: workaround for layout bug ( Qt < 4.3 )
   of the Cleanlook style added.
7. A couple of minor fixes

Changes
-------
1. QSvgGenerator added to the bode example

Release 5.0.1
===================

Changes
-------
1.  A couple of problems, when building Qwt fixed.
2.  Displaying Rich Text with Qt 4.x fixed

Release 5.0.0
===================

Platforms
---------
Support of Qt3 and Qt4. Qt2 is not supported any longer.

Key features
------------
1.  Redesign of plot items. Makes it much easier to develop
    individual items.
2.  Redesign of the scale classes. All calculations are
    collected in scale engines, where the application can
    implement it´s own (f.e log2, or date scales). Now it´s
    also possible to have individual and completely irregular scales
3.  Redesign of the QwtText classes. The MathML renderer of
    the Qt4 solutions package is embedded. 
    work for all expressions/situations.
4.  New classes for navigating: QwtPanner, QwtMaginfier
5.  Spectrogram/Contour plots  and other classes for displaying
    raster data added.

Changes
-------
5.0.0 is by far the release with the most changes in the history of Qwt
- too many to make list.


Release 4.2.0/0.4.2
===================

License
--------
A couple of exceptions to the LGPL with the intention to allow static
linking with commercial applications. See COPYING.

Key features: 
-------------
1.  Designer plugin 
2.  Rich Text support ( f.e. E = m * c<sup>2</sup> ) added.
3.  QwtDial class family added (QwtDial, QwtCompass, QwtAnalogClock, ...)
4.  QwtPicker class family added. Includes QwtPlotZoomer, a complete
    implementation of recursive zooming.
5.  Device metrics independent printing of QwtPlot. (QPrinter::HighResolution)
6.  QwtPlot::setCurveBrush(), QwtCurve::setBrush() added. The area
    between curve and baseline will be filled with this brush.
7.  Rotation of axis tick labels added. Very useful for axis with long
    labels like time scales ...
8.  Added a new abstract QwtData class to plot data from almost any type of
    container class.  
9.  QwtDoublePoint, QwtDoubleSize, QwtDoubleRect double counterparts
    for QPoint, QSize, QRect.
10. First steps to support Qtopia. All examples can be compiled and started
    in the qvfb emulator.


Changes:
---------
1.  Rewrite of QwtLegend/QwtLegendItem (no QTable anymore)
2.  Each plot item will be painted, even if one of the axis it is 
    attached to is disabled. (like in all other releases beside 0.4.1)
3.  Code for double buffering moved to a new class QwtPaintBuffer.
    Double buffering can be enabled/disabled now. 
4.  QwtPainter, QwtMetricsMap, QwtLayoutMetrics added
    Hide paint device metrics dependencies.
5.  Layout code rewritten and moved to a new class QwtPlotLayout
    New layout options canvasMargin(), alignCanvasToScales()
6.  QwtPlot: sizeHint() != minimumSizeHint()
9.  Internal plot data are private again. A couple of get methods 
    added instead.
10. canvas repaints triggered by paint events. Enables event filtering
11. QwtPlot::drawCanvasItems added. In opposite to QwtPlot::drawCanvas
    it is used by the printing code too.
12. qwtMax, qwtMin, qwtInt mapped to QMAX, QMIN, qRound from qglobal.h
13. operator= for plot item classes changed.
14. readOnly property added for sliders.
15. valid flag added for QwtDblRange
16. QwtCounter wrap around policy: a counter under- or overflow sets
    focus to the smallest up/down button and disables counting.
    A space bar keypress release event re-enables counting.
17. QwtPushButton added. A class that adds rich text and alignments
    features to QPushButton, like they are used in QLabel
18. Clipped painting code moved from QwtCurve to QwtPainter/QwtRect
19. Canvas cache added to optimize trivial repaints.
20. QwtPlot::drawCurve added for incremental curve data
21. QwtSliderBase, readOnly, isValid added
22. Added filtering of the colors of the title and scales to QwtPrintFilter.
23. Support of QT_NO_CAST_ASII and QT_NO_COMPAT added

libsmi/tools/smiquery.c
/*
 * smiquery.c --
 *
 *      A simple SMI query program, using libsmi.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smiquery.c 1797 2013-06-06 09:13:00Z schoenw $
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"

static const int indent = 12;


static void
print_key(const char *s)
{
    if (s) {
	printf("%*s: ", indent, s);
    } else {
	printf("%*s  ", indent, "");
    }
}


static void
print_val(const char *s)
{
    const char *p;
    
    if (!s) {
	printf("-");
    } else {
	for (p = s; *p; p++) {
	    putc(*p, stdout);
	    if (*p == '\n') {
		print_key(NULL);
	    }
	}
    }
    putc('\n', stdout);
}


static void
print_key_val(const char *key, const char *val)
{
    print_key(key);
    print_val(val);
}


static void usage()
{
    fprintf(stderr,
	    "Usage: smiquery [options] command name [command name ...]\n"
	    "  -V, --version        show version and license information\n"
	    "  -h, --help           show usage information\n"
	    "  -c, --config=file    load a specific configuration file\n"
	    "  -p, --preload=module preload <module>\n"
	    "\nSupported commands are:\n"
	    "  module <module>           show information on module <module>\n"
	    "  imports <module>          show import list of module <module>\n"
	    "  node <module::name>       show information on node <module::name>\n"
	    "  compliance <module::name> show information on compliance node <module::name>\n"
	    "  children <module::name>   show children list of node <module::name>\n"
	    "  type <module::name>       show information on type <module::name>\n"
	    "  macro <module::name>      show information on macro <module::name>\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smiquery " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smiquery"); }
static void preload(char *module) { smiLoadModule(module); }



int main(int argc, char *argv[])
{
    SmiModule *module;
    SmiNode *node, *node2, *child, *parent;
    SmiType *type, *parenttype;
    SmiMacro *macro;
    SmiNamedNumber *nn;
    SmiRange *range;
    SmiImport *import;
    SmiRevision *revision;
    SmiOption *option;
    SmiRefinement *refinement;
    SmiElement *element;
    char *command, *name, *p;
    int i;
    char s1[40], s2[40];

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smiquery");
    else
	smiInit(NULL);

    optParseOptions(&argc, argv, opt, 0);

    if (!(argc % 2)) {
	usage();
	return 1;
    }

    for (i = 1; i < argc; i +=2) {

	command = argv[i];
	name = argv[i+1];

	if (!strcmp(command, "module")) {
	    module = smiGetModule(name);
	    if (module) {
		node = smiGetModuleIdentityNode(module);
		print_key_val("Module", module->name);
		if (module->path)
		    print_key_val("Pathname", module->path);
		if (node)
		    print_key_val("Object",
				  smiRenderNode(node, SMI_RENDER_ALL));
		if (module->organization)
		    print_key_val("Organization", module->organization);
		if (module->contactinfo)
		    print_key_val("ContactInfo", module->contactinfo);
		if (module->description)
		    print_key_val("Description", module->description);
		if (module->reference)
		    print_key_val("Reference", module->reference);
		print_key_val("Language",
			      smiLanguageAsString(module->language));
		print_key("Conformance"); printf("%d\n", module->conformance);
		print_key_val("Loaded", smiIsLoaded(name) ? "yes" : "no");

		for(revision = smiGetFirstRevision(module);
		    revision ; revision = smiGetNextRevision(revision)) {
		    print_key_val("Revision", ctime(&revision->date));
		    if (revision->description)
			print_key_val("Description", revision->description);
		}
	    }
	} else if (!strcmp(command, "imports")) {
	    module = smiGetModule(name);
	    if (module && smiGetFirstImport(module)) {
		print_key("Imports");
		for(import = smiGetFirstImport(module); import; ) {
		    printf("%s::%s\n", import->module, import->name);
		    import = smiGetNextImport(import);
		    if (import) {
			print_key(NULL);
		    }
		}
	    }
	} else if (!strcmp(command, "node")) {
	    node = smiGetNode(NULL, name);
	    if (node) {
		parent = smiGetParentNode(node);
	    }
	    if (node) {
		type = smiGetNodeType(node);
		print_key_val("MibNode", smiRenderNode(node, SMI_RENDER_ALL));
		print_key_val("OID", smiRenderOID(node->oidlen, node->oid, 0));
		for (node2 = smiGetFirstAlias(node);
		     node2; node2 = smiGetNextAlias(node2)) {
		    if (node2 == node) continue;
		    print_key_val("Alias",
				  smiRenderNode(node2, SMI_RENDER_ALL));
		}
		
		if (parent)
		    print_key_val("ParentNode",
				  smiRenderNode(parent, SMI_RENDER_ALL));
		if (type)
		    print_key_val("Type", smiRenderType(type, SMI_RENDER_ALL));
		if (node->value.basetype != SMI_BASETYPE_UNKNOWN)
		    print_key_val("Default",
			  smiRenderValue(&node->value, type, SMI_RENDER_ALL));
		if (node->decl != SMI_DECL_UNKNOWN)
		    print_key_val("Declaration", smiDeclAsString(node->decl));
		print_key_val("NodeKind", smiNodekindAsString(node->nodekind));
		if (node->nodekind == SMI_NODEKIND_ROW) {
		    print_key_val("Creatable", node->create ? "yes" : "no");
		    print_key_val("Implied", node->implied ? "yes" : "no");
		}
		switch (node->nodekind) {
		case SMI_NODEKIND_ROW: p = "Index"; break;
		case SMI_NODEKIND_COMPLIANCE: p = "Mandatory"; break;
		case SMI_NODEKIND_CAPABILITIES: p = "Includes"; break;
		case SMI_NODEKIND_GROUP: p = "Members"; break;
		case SMI_NODEKIND_NOTIFICATION: p = "Objects"; break;
		default: p = "Elements";
		}
		if (smiGetFirstElement(node)) {
		    print_key(p);
		    for(element = smiGetFirstElement(node);
			element; ) {
			node2 = smiGetElementNode(element);
			print_val(smiRenderNode(node2, SMI_RENDER_ALL));
			element = smiGetNextElement(element);
			if (element) {
			    print_key(NULL);
			}
		    }
		}
		if (node->access != SMI_ACCESS_UNKNOWN)
		    print_key_val("Access", smiAccessAsString(node->access));
		if (node->status != SMI_STATUS_UNKNOWN)
		    print_key_val("Status", smiStatusAsString(node->status));
		if (node->format)
		    print_key_val("Format", node->format);
		if (node->units)
		    print_key_val("Units", node->units);
		if (node->description)
		    print_key_val("Description", node->description);
		if (node->reference)
		    print_key_val("Reference", node->reference);
	    }
	} else if (!strcmp(command, "compliance")) {
	    node = smiGetNode(NULL, name);
	    if (node) {
		if (smiGetFirstElement(node)) {
		    print_key("Mandatory");
		    for(element = smiGetFirstElement(node);
			element ; ) {
			node2 = smiGetElementNode(element);
			print_val(smiRenderNode(node2, SMI_RENDER_ALL));
			element = smiGetNextElement(element);
			if (element) {
			    print_key(NULL);
			}
		    }
		}
		if (smiGetFirstOption(node)) {
		    for(option = smiGetFirstOption(node); option ;) {
			node2 = smiGetOptionNode(option);
			print_key_val("Option",
				      smiRenderNode(node2, SMI_RENDER_ALL));
			if (option->description)
			    print_key_val("Description",
					  option->description);
			option = smiGetNextOption(option);
		    }
		}
		if (smiGetFirstRefinement(node)) {
		    for(refinement = smiGetFirstRefinement(node);
			refinement ;
			refinement = smiGetNextRefinement(refinement)) {
			node2 = smiGetRefinementNode(refinement);
			print_key_val("Refinement",
				      smiRenderNode(node2, SMI_RENDER_ALL));
			type = smiGetRefinementType(refinement);
			if (type) {
			    print_key_val("Type",
				  smiRenderType(type, SMI_RENDER_ALL));
			}
			type = smiGetRefinementWriteType(refinement);
			if (type) {
			    print_key_val("Write-Type",
				  smiRenderType(type, SMI_RENDER_ALL));
			}
			if (refinement->access != SMI_ACCESS_UNKNOWN) {
			    print_key_val("Access",
				  smiAccessAsString(refinement->access));
			}
			if (refinement->description)
			    print_key_val("Description",
					  refinement->description);
		    }
		}
	    }
	} else if (!strcmp(command, "children")) {
	    node = smiGetNode(NULL, name);
	    if (node && smiGetFirstChildNode(node)) {
		print_key("Children");
		for(child = smiGetFirstChildNode(node); child ; ) {
		    print_val(smiRenderNode(child, SMI_RENDER_ALL));
		    child = smiGetNextChildNode(child);
		    if (child) {
			print_key(NULL);
		    }
		}
	    }
	} else if (!strcmp(command, "type")) {
	    p = strrchr(name, ':');
	    if (!p) p = strrchr(name, '.');
	    if (!p) p = strrchr(name, '!');
	    if (p) {
		p++;
	    } else {
		p = name;
	    }
	    if (islower((int)name[0]) || isdigit((int)name[0]) ||
		!isupper((int)p[0])) {
		node = smiGetNode(NULL, name);
		if (node) {
		    type = smiGetNodeType(node);
		} else {
		    type = smiGetType(NULL, name);
		}
	    } else {
		type = smiGetType(NULL, name);
	    }
	    if (type) {
		parenttype = smiGetParentType(type);
		print_key_val("Type", smiRenderType(type, SMI_RENDER_ALL));
		print_key_val("Basetype", smiBasetypeAsString(type->basetype));
		if (parenttype)
		    print_key_val("Parent Type",
				  smiRenderType(parenttype, SMI_RENDER_ALL));
		if (type->value.basetype != SMI_BASETYPE_UNKNOWN)
		    print_key_val("Default",
			  smiRenderValue(&type->value, type, SMI_RENDER_ALL));
		if ((type->basetype == SMI_BASETYPE_ENUM) ||
		    (type->basetype == SMI_BASETYPE_BITS)) {
		    if (smiGetFirstNamedNumber(type)) {
			print_key("Numbers");
			for(nn = smiGetFirstNamedNumber(type); nn; ) {
			    printf("%s(%ld)\n",
				   nn->name, nn->value.value.integer32);
			    nn = smiGetNextNamedNumber(nn);
			    if (nn) {
				print_key(NULL);
			    }
			}
		    }
		} else {
		    if (smiGetFirstRange(type)) {
			print_key("Ranges");
			for(range = smiGetFirstRange(type); range; ) {
			    strcpy(s1, smiRenderValue(&range->minValue, type, 0));
			    strcpy(s2, smiRenderValue(&range->maxValue, type, 0));
			    printf("%s", s1);
			    if (strcmp(s1, s2)) printf("..%s", s2);
			    printf("\n");
			    range = smiGetNextRange(range);
			    if (range) {
				print_key(NULL);
			    }
			}
		    }
		}
		print_key_val("Declaration", smiDeclAsString(type->decl));
		if (type->status != SMI_STATUS_UNKNOWN)
		    print_key_val("Status", smiStatusAsString(type->status));
		if (type->format)
		    print_key_val("Format", type->format);
		if (type->units)
		    print_key_val("Units", type->units);
		if (type->description)
		    print_key_val("Description", type->description);
		if (type->reference)
		    print_key_val("Reference", type->reference);
	    }
	} else if (!strcmp(command, "macro")) {
	    macro = smiGetMacro(NULL, name);
	    if (macro) {
		print_key_val("Macro", macro->name);
		if (macro->status != SMI_STATUS_UNKNOWN)
		    print_key_val("Status", smiStatusAsString(macro->status));
		if (macro->description)
		    print_key_val("Description", macro->description);
		if (macro->reference)
		    print_key_val("Reference", macro->reference);
	    }
	} else {
	    usage();
	    return 1;
	}
    }
    
    smiExit();
    
    return 0;
}
